Use SQLCipher to encrypt SQLite databases: datastores & indexes (second try)

indisoluble opened the pull request:
*What:*
Provide a mechanism to cipher the SQLite databases in CDTDatastore with SQLCipher.

*Why:*
Some users can not rely on the encryption provided by default in iOS. We have to give an alternative solution for those that are in this situation.

*How:*
CDTDatastore accesses to SQLite databases with FMDB. This library is already able to use SQLCipher, we only have to provide a key to encrypt the database and compile the library with the correct subspec. However users should not have to care about 3rd party libraries so:
* We will add a subspec to CDTDatastore.podspec that will import every necessary component to encrypt databases.
* Only if the user set this new subspec, CDTDatastore will expose the new methods to create encrypted datastores. We will handle this defining the new methods in categories, only with the new subspec these categories will be imported into the project. This way, if an user does not change his/her Podfile, the API will remain the same and he/she will not be able to use the new methods by mistake.
* To provide the key to encrypt the databases, a user will have to implement a class that conforms to a new protocol and pass an instance of the class to the new methods. Defining a protocol will allow us to implement later on a helper class to safely handle and store the key.

*Tests:*
CDTDatastore manages two different databases:
- One with the data in the datastore
- Other for the indexes

Each one is tested independently. Also the result of these tests depends on the library included, i.e. if SQLCipher is included or not.

* datastore + SQLCipher
  1. Create a non-encrypted db if no key is provided
  2. Create an encrypted db if a key is provided
  3. Do not open a non-encrypted db if a key is provided
  4. Do not open an encrypted db if no key is provided
  5. Do not open an encrypted db if the key provided is not the key used to encrypt the db
  6. Do not open a non-encrypted db if a key is provided although the db was opened before (in the same session)
  7. Do not open an encrypted db if no key is provided although the db was opened before (in the same session)
  8. Do not open an encrypted db if the key provided is not the key used to encrypt the db although the db was opened before (in the same session)

* datastore + SQLite
  1. Create a non-encrypted db if no key is provided
  2. Do not create a encrypted db if a key is provided
  3. Do not open a non-encrypted db with a key
  4. Do not open an encrypted db with a key (although it is the right key)
  5. Do not open an encrypted db if no key is provided
  3. Do not open a non-encrypted db with a key although the db was opened before (in the same session)

* index + SQLCipher
  1. Create an non-encrypted db with a non-encrypted datastore
  2. Create an encrypted db with an encrypted datastore
  3. Do not open an encrypted db with a non-encrypted datastore
  4. Do not open a non-encrypted db with an encrypted datastore
  5. Do not open an encrypted db with an encrypted datastore if they do not share the same key

* index + SQLite
  1. Create an non-encrypted db with a non-encrypted datastore
  2. Do not open an encrypted db with a non-encrypted datastore


reviewer @mikerhodes
reviewer @emlaver

mikerhodes commented:
Main things:

- key _provider_ not _retriever_. The implementor of the protocol need not retrieve a key from anywhere (e.g., your hard-coded implementations in the tests).
- Your tests don't appear to check that the database file is actually encrypted. Given you're providing a fixed key, you should be able to check the first say 50 bytes of the file against a canned set of bytes.
- 09b74c2 has a bunch of stuff related to deleting databases in it. Should these be in the other commit?
- I don't think that the createDatabase=true option in the replicator should allow creating databases with encryption keys. Extra complexity isn't worth it.
- I'd like to understand why the pattern of a "dummy" key provider rather than just allowing a `nil` provider?

indisoluble commented:
Changed:
* Protocol renamed: key provider not retriever.
* Tests check that the database (datastore and index) is encrypted or no depending on the case.
* Replicators do not use (or make reference) to encryption key. They are created with already open databases, so they do not need the encryption key.

Not changed:
* It was necessary to modify the delete method because it opened the databases before deleting them, therefore it needed the encryption key. To avoid that, I check if the database is already open because in the case I do not need the encryption key and if it is not open, no other object is related to it so I can simply delete the files from disk.
* I do not allow nil as a provider because in that case I have to do twice validations. First to check if there is an object and second to check if I get an encryption key. Of course, if I send a method to nil, I'll get nil as a response but if I try to put the provider in an array or a dictionary, the app will crash.
* Some classes implement NSCopying protocol but they do not copy the object, instead they alloc a new object; that is because they inherit from NSObject.

Also:
* Renamed *dummy provider* to *nil provider*.
* Methods in CDT-classes related to encrypt the database have been moved to categories, so if a user is not interested in this functionality, he/she does not have to bother about it.
* Clarified code in 'TD_DatabaseManager:deleteDatabaseNamed:error:'

Pending:
* Rewrite history before reviewing changes.

mikerhodes commented on Classes/common/CDTEncryptionKey/CDTDatastore+EncryptionKey.h:
```diff
@@ -0,0 +1,39 @@
+
+#import "CDTDatastore.h"
+
+@protocol CDTEncryptionKeyProviding;
```
I think this should be `CDTEncryptionKeyProvider`

indisoluble replied:
Fixed in 66c19fb

indisoluble commented:
**Test cases**

CDTDatastore manages two different databases:
- One with the data in the datastore
- Other for the indexes

Each one is tested independently. Also the result of these tests depends on the library included, i.e. if SQLCipher is included or not.

* datastore + SQLCipher
  1. Create a non-encrypted db if no key is provided
  2. Create an encrypted db if a key is provided
  3. Do not open a non-encrypted db if a key is provided
  4. Do not open an encrypted db if no key is provided
  5. Do not open an encrypted db if the key provided is not the key used to encrypt the db
  6. Do not open a non-encrypted db if a key is provided although the db was opened before (in the same session)
  7. Do not open an encrypted db if no key is provided although the db was opened before (in the same session)
  8. Do not open an encrypted db if the key provided is not the key used to encrypt the db although the db was opened before (in the same session)

* datastore + SQLite
  1. Create a non-encrypted db if no key is provided
  2. Fails if it tries to create/open a db with a key
  3. Fails if it tries to re-open a db with a key

* index + SQLCipher
  1. Create an non-encrypted db with a non-encrypted datastore
  2. Create an encrypted db with an encrypted datastore

* index + SQLite
  1. Create an non-encrypted db with a non-encrypted datastore

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -76,18 +79,46 @@ static BOOL removeItemIfExists(NSString* path, NSError** outError)
+    NSString *str = [NSString stringWithCString:buffer encoding:NSASCIIStringEncoding];
+    
+    return ![TD_DatabaseStandardSQLiteHeader isEqualToString:str];
+}
```
This method actually checks whether there is a not-unencrypted database at `path` -- it could just as well be any file other than one with a normal SQLite header.

I'd flip the logic of the method and name it `-isDatabaseUnencryptedAtPath:` as that's the only thing you can prove without attempting to decrypt the database.

In addition, could you just read the first `[TD_DatabaseStandardSQLiteHeader length]` bytes from the file rather than the whole thing into an NSData?

indisoluble replied:
Fixed in 6d4e47a

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -193,47 +229,140 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
+        } else {
+            flags |= SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE;
+        }
+        CDTLogInfo(CDTDATASTORE_LOG_CONTEXT, @"Open %@ (flags=%X)", _path, flags);
```
`Info` seems rather high a level for this; suggest `Debug`.

indisoluble replied:
Fixed in fbb2609

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -76,18 +79,46 @@ static BOOL removeItemIfExists(NSString* path, NSError** outError)
+    NSFileHandle *fileHandle = [NSFileHandle fileHandleForReadingAtPath:path];
+    if (!fileHandle) {
+        return YES;
+    }
```
What's your reasoning for returning `YES` here?

indisoluble replied:
At the moment I wrote that `if`, it looked like the right thing; if the db does not exist, then it is not encrypted but likewise we can state the contrary, both options are right. However returning `YES` is the right answer for the code in this [line](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-8e8b4e1d26db1376fbf88a600a1a7a1fR242).

Probably, it would be a better idea to return an `enum` value:
- isUnencrypted
- isEncrypted
- fileNotFound

What do you think?

indisoluble replied:
Fixed in 05ef2e4

mikerhodes commented:
Several of the error and other log messages in `-openFMDBWithEncryptionKeyProvider:` need more context to them, for example not all include the path or are difficult to understand out of context, such as `@"Wrong key"`. At least include the database name.

Also remember log messages are read by a developer, so messages like `Is DB at path %@ encrypted?` are read as questions to them, where as the log message's meaning is more like "You didn't provide a key. Checking whether database is encrypted.".

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
+        result = (queue != nil);
+    }
+
+    // Cipher database
```
Comment seems inaccurate -- this checks we can use the key provided with the database file?

indisoluble replied:
Fixed in bc0c17e

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
+    
+    if ([self isOpen]) {
+         return [self isOpenWithEncryptionKeyProvider:provider];
+    }
```
Is this return value right? It seems like you can return `NO` while the database is actually open already.

It depends what you see as the contract for the method here:

a. Try to open the database _with this exact key_.
b. Try to open the database, _here's a key you can try_.

For (b), a simple `return YES` would suffice here, and your method works for (a).

What's the intent here, and can we make it clearer?

indisoluble replied:
The intent for method `openWithEncryptionKeyProvider:` is (b).

I think the name already clarifies the purpose because it is not the only method that uses the suffix *WithEncryptionKeyProvider:*, `isOpenWithEncryptionKeyProvider:` also uses it and has the same meaning. Next to this method there is another one without the suffix called `isOpen` that simply verifies if the TD_Database instance is opened with any key.

Also, all these three methods are commented in `TD_Database.h`: [1](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-d0d702218592f847785da5ec2f8d4e99R94), [2](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-d0d702218592f847785da5ec2f8d4e99R83), [3](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-d0d702218592f847785da5ec2f8d4e99R78) and the comments explain how the keyProvider is used.

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
+    return _open;
+}
+
+// callers: many things
```
According to Xcode, there's actually only one caller of this method.

indisoluble replied:
Yes but it is a public method, potentially it can be called from many places. Right now only `openWithEncryptionKeyProvider:` calls it but I think we can need it later on so I made it public.

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
 {
-    if (_open) return YES;
-    if (![self openFMDB]) return NO;
+    Assert(provider, @"Key provider is mandatory. Inform a nil provider instead");
```
"Supply a CDTNilEncryptionKeyProvider instead."

indisoluble replied:
Fixed in 1b603f0

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -438,24 +577,43 @@ - (BOOL)close
-    } else if (!self.exists) {
-        return YES;
+
+    if (_open && ![self close]) {
```
Is it possible here for `_open` to be `YES`, then this thread is interrupted and the database is closed, in which case `[self close]` will return `NO`? If that can happen, we end up failing this method even though the database is closed and we could go on to delete it.

Even if the above is the case, it may not be important enough to fix, given the method could be called again fairly simply -- particularly if you fill in the `outError` appropriately.

indisoluble replied:
`[self close]`can not fail, it only returns `NO` if the db is not open and we already validates that before calling it. I wrote the code that way to keep it similar to how it was before but it does not make sense, I can simplify it to:

```
    [self close];
```

mikerhodes replied:
Presuming the code can be called from multiple threads (I've not checked this, but it doesn't look like there's anything preventing this), the database could be closed by another thread between the execution of the check on `_open` and the call to `close`. I'm wondering:

a. Whether this can happen;
b. Whether its worth fixing (it doesn't look like there is potential for data loss, for example).

indisoluble replied:
As a resume, the main concern here is that this code is not thread-safe. However, we have this problem not only when we 'delete' a db, we also have the same issue when we 'open' and 'close' a db.

Once a db is created and opened, all operations are executed in a serial FMDBQueue but the operations mentioned before are executed in the same thread they are called and there are not locks in place to manage concurrency.

To end, this is a bug and we will address it in a new PR.

mikerhodes commented on Classes/common/touchdb/TD_Database.m:
```diff
@@ -438,24 +577,43 @@ - (BOOL)close
+        NSString *attachmentsPath = [TD_Database attachmentStorePathWithDatabasePath:path];
+
+        success =
+            (removeItemIfExists(path, outError) && removeItemIfExists(attachmentsPath, outError));
```
A crash after `removeItemIfExists(path, outError)` would leave the attachments data hanging around and one would never be able to delete it through this method because the conditional would fail because the database would be gone.

indisoluble replied:
In that case, should I try to delete first the attachments? If it crashes, I'll still be able to delete the database if I try again.

mikerhodes replied:
Personally, I'm not sure the `if` check is needed, given that the method name `removeItemIfExists` sounds like it'll do something similar anyway, and anyway something could easily delete the database between the `existsDatabaseAtPath:path` call and the `removeItemIfExists` call it appears to be attempting to protect.

It's important to remember that a lot of this code assumed single threading. While I corrected most when I moved to GCD, we should probably be cleaning up code like this that relies on those assumptions as we go.

indisoluble replied:
Check previous [comment](https://github.com/cloudant/CDTDatastore/pull/112#discussion_r27375424).

As we already mentioned, we have a concurrency issue when it comes to 'open', 'close' & 'delete' operations.

We will address this problem with a new PR.

mikerhodes commented on Classes/common/CloudantSync.h:
```diff
@@ -15,7 +15,11 @@
+#import "CDTDatastoreManager+EncryptionKey.h"
+#else
 #import "CDTDatastoreManager.h"
+#endif
```
Probably worth a small comment explaining this, and where ENCRYPT_DATABASE is set.

indisoluble replied:
Fixed in 996fd3a

mikerhodes commented:
I think you could split up 9127d102fad8883287102de6d6f0800ae03f7c44, it's over 3000 lines in a single commit. Could you have something like:

1. `Update podspec to allow building with encryption support` -- this would allow you to explain how the subspecs work in the commit comment.
2. `Allow user to provide key to encrypt JSON data` -- which would just be the code the get the encryption enabled on the main database with the JSON docs in.
3. `Encrypt database indexes` -- which shows how the key is carried into the indexing code (this would be particularly useful when updating Cloudant Query to support this after the merge).
4. `Add tests to confirm databases encrypted` -- adding the new project and tests (which is a large number of lines for not that many changes, and would be helpful separated).

mikerhodes commented on Classes/common/Indexing/CDTIndexManager.m:
```diff
@@ -823,6 +759,132 @@ - (BOOL)isValidIndexName:(NSString *)indexName error:(NSError *__autoreleasing *
+    }
+
+    return success;
+}
```
This has far fewer checks than the version inside TD_Database. E.g.:

- No check for whether the database is encrypted if there is no key provided.
- You don't do the query to `sqlite_master` to make sure the key actually works.

It might be worth pulling out a utility class which does all these checks so both TD_Database and CDTIndexManager (and the Query index manager) can use it.

indisoluble replied:
But an index manager is created with an datastore. If the datastore is correctly initialised, this code will not fail. It will fail if the datastore is encrypted but the index is not, which means that the user copied in the folder an index created with a different datastore. Do we need to add defensive code for a situation like this?

mikerhodes replied:
I don't think you should assume that the data on disk is consistent with each other. As you say, it's easy enough for an index database to be copied in whose encryption or lack thereof doesn't match the one on the main database.

In addition, in making the assumption that the TD_Database object does all the checks you need, you're increasing coupling between the two objects and the knowledge that someone maintaining the code needs to have of the wider system.

indisoluble replied:
Fixed in dc84c4b

mikerhodes commented on Classes/common/CDTEncryptionKey/CDTEncryptionKeyProvider.h:
```diff
@@ -0,0 +1,35 @@
+//
```
Suggest naming the folder/group "Encryption".

indisoluble replied:
Fixed in e592f57

mikerhodes commented on Classes/common/CDTEncryptionKey/CDTEncryptionKeyNilProvider.h:
```diff
@@ -0,0 +1,25 @@
+
+#import "CDTEncryptionKeyProvider.h"
+
+@interface CDTEncryptionKeyNilProvider : NSObject <CDTEncryptionKeyProvider>
```
Needs documentation as you expect external developers to use this (it turns up in your error messages).

indisoluble replied:
Fixed in 6633e73

mikerhodes commented on Classes/common/CDTEncryptionKey/CDTDatastoreManager+EncryptionKey.h:
```diff
@@ -0,0 +1,45 @@
+ saving to disk (attachments and extensions not included).
+ If a key is provided the first time the datastore is open, only this key will be valid the next
+ time. In the same way, if no key is informed, the datastore will not be cipher and can not be
+ cipher later on.
```
This comment isn't accurate -- you need to discuss the provider and it functionality (e.g., from what I recall `provider` cannot be `nil`).

indisoluble replied:
Fixed in ad3debd

mikerhodes commented on Classes/common/CDTEncryptionKey/CDTDatastoreManager+EncryptionKey.h:
```diff
@@ -0,0 +1,45 @@
+ @return a datastore for the given name
+
+ @warning *Warning:* Encryption is an experimental feature, use with caution. It won't work unless
+ you use subspec 'CDTDatastore/SQLCipher'
```
Is encryption experimental? I was hoping not, though the warning to use `CDTDatastore/SQLCipher` is useful.

indisoluble replied:
No, it is not experimental but at the moment I wrote this comment I considered this development part of a bigger one so it was experimental in the sense that there were still more changes to deploy.

indisoluble replied:
Fixed in ad3debd

mikerhodes commented on Classes/common/CDTEncryptionKey/CDTEncryptionKeyNilProvider.m:
```diff
@@ -0,0 +1,36 @@
+- (BOOL)isEqual:(id)object
+{
+    return (object && [object isMemberOfClass:[CDTEncryptionKeyNilProvider class]]);
+}
```
Do you need to override `-hash` here too?

indisoluble replied:
The problem here is that I do not need `isEqual:`, as far as I can remember I overrided this method to perform some test that I discarded later on.

mikerhodes replied:
I guess the solution here is to remove `-isEqual`.

indisoluble replied:
Fixed in 98dc44c

mikerhodes commented on Classes/common/CDTDatastore.m:
```diff
@@ -52,23 +55,38 @@ + (NSString *)versionString { return @CLOUDANT_SYNC_VERSION; }
-        _database = database;
-        if (![_database open]) {
-            return nil;
+        id<CDTEncryptionKeyProvider> otherProvider = [provider copy];
```
Not a good variable name -- it's a copy, not a different provider object.

indisoluble replied:
Fixed in 9775655

mikerhodes commented on Classes/common/CDTDatastore.m:
```diff
@@ -150,6 +168,11 @@ - (NSUInteger)documentCount
+- (id<CDTEncryptionKeyProvider>)copyEncryptionKeyProvider
+{
+    return [_encryptionKeyProvider copy];
+}
```
Could this just be `encryptionKeyProvider`? Does the caller need to know it's a copy?

indisoluble replied:
When a datastore is initialised, [I copy the key provider passed as a parameter](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-09611c62c3ef826c53341a9070bc3d7dR70). I do that to ensure that the key does not change once the datastore is created. For this same reason, I have to return a copy and not the key provider stored as a property.

Given that I return a copy (retain count +1), I have to prefix the method with `copy` to follow the [naming convention](http://clang.llvm.org/docs/AutomaticReferenceCounting.html#method-families).

mikerhodes replied:
Copying the key provider doesn't provide the guarantee you describe because the provider can get its key from elsewhere in a dynamic manner (e.g., network request). I'm therefore unsure that this precaution gains us something here, and ends up being a little confusing.

indisoluble replied:
Ok, in that case it is not worth to copy the key provider at all, even during the initialisation of a datastore. I'll remove the `NSCopying` from `CDTEncryptionKeyProvider` as well as the method `copy`.

indisoluble replied:
Fixed in 9775655

mikerhodes commented on CONTRIBUTING.md:
```diff
@@ -101,6 +107,44 @@ shortcut.
+access SQLite databases and [FMDB][FMDB] is able to do that using the standard
+iOS library for SQLite or [SQLCipher][SQLCipher].
+
+If [SQLCipher][SQLCipher] is not included but you try to create a `CDTDatastore`
```
Only the first instance needs to be hyperlinked.

indisoluble replied:
Fixed in b34babf

mikerhodes commented on CONTRIBUTING.md:
```diff
@@ -101,6 +107,44 @@ shortcut.
+however the behaviour is different depending of the libraries included. The
+`Podfile` in `EncryptionTests` is configured to import [FMDB][FMDB] with
+[SQLCipher][SQLCipher] while `CDTDatastore.xcworkspace` uses the default version
+of [FMDB][FMDB]. `EncryptionTests.xcworkspace` has a sub-set of the tests in
```
Again, only hyperlink the first instance.

indisoluble replied:
Fixed in b34babf

mikerhodes commented on CDTDatastore.podspec:
```diff
@@ -29,27 +29,53 @@ Pod::Spec.new do |s|
+
+      ssp.dependency 'FMDB/SQLCipher', '= 2.3'
+      ssp.xcconfig = { 'OTHER_CFLAGS' => '$(inherited) -DENCRYPT_DATABASE' }
+    end
```
Does this not work if it's outside the `common` subspec, instead being just inside the top-level `SQLCipher` subspec?

indisoluble replied:
... I'm afraid there is a problem here. I re-checked the log generated by `pod lib lint` after reading this comment and I found the following when the subspec `CDTDatastore/SQLCipher` is tested:

```
...
            Requiring nested dependencies ()
            Creating possibility state for CDTDatastore/common/SQLCipher (= 0.14.0) (1 remaining)
              Attempting to activate CDTDatastore/common/SQLCipher (0.14.0)
              Activated CDTDatastore/common/SQLCipher at CDTDatastore/common/SQLCipher (0.14.0)
              Requiring nested dependencies (CDTDatastore/no-arc (= 0.14.0), CocoaLumberjack (= 2.0.0-rc2), FMDB/SQLCipher (= 2.3))
              Creating possibility state for CocoaLumberjack (= 2.0.0-rc2) (1 remaining)
                Attempting to activate CocoaLumberjack (2.0.0-rc2)
                Found existing spec (CocoaLumberjack (2.0.0-rc2))
                Creating possibility state for CDTDatastore/common/SQLite (= 0.14.0) (1 remaining)
                  Attempting to activate CDTDatastore/common/SQLite (0.14.0)
                  Activated CDTDatastore/common/SQLite at CDTDatastore/common/SQLite (0.14.0)
...
```

It looks like it is importing both subspecs, the one with the standard SQLite library and the one with SQLCipher. That can not be right.

indisoluble replied:
There is no easy solution for this problem.

Essentially `pod lib lint` checks that the subspecs compile, i.e:

- [SQLite](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-06ec762bd04c5ee54e244d1e3fd0d111R53) depends only on **FMDB**, after downloading this source code, `pod lib lint` tries to compile it and if it succeeds, the validation passes.
- [common](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-06ec762bd04c5ee54e244d1e3fd0d111R37) depends on [no-arc](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-06ec762bd04c5ee54e244d1e3fd0d111R67) and **CocoaLumberjack** and the default subspec is `SQLite`. `pod lib lint` downloads and copies the necessary source code and then compiles; again if it succeeds, the validation passes.

Now, our CDT-code depends on **FMDB** and **FMDB** depends on **SQLite** or **SQLCipher**. `pod lib lint` will not pass if I define a subspec called `common` with our code but without **FMDB**. If I define 2 subspecs: one that imports `FMDB` called `SQLite` and other that imports `FMDB/SQLCipher` called `SQLCipher`, and then I add as a dependency `SQLite` to `common`, `pod lib lint` will succeed but `SQLCipher` is worthless because if someone uses it in his/her `Podfile`: `pod CDTDatastore/SQLCipher`, the only thing he/she will include in his/her workspace will be **FMDB** with **SQLCipher**. Then if I add `common` to `SQLCipher`, `pod lib lint` will succeed but I'll have again the problem I was trying to avoid (I'm importing both **SQLite** and **SQLCipher**).

So:
- Our code can not be compiled in isolation
- It requires `FMDB` or `FMDB/SQLCipher`
- But I do not want to compile the code with both dependencies added at the same time
- Also I want to provide to the user one and only one subspec that, once included in his/her `Podfile`, will add to the workspace all the required source code: **FMDB**, **SQLCipher** & **CDTDatastore**. If he/she types `pod CDTDatastore`, only **FMDB** & **CDTDatastore** will be included.

I can only think of 3 possible solutions:
- Remove `pod lib lint --allow-warnings` from `Rakefile`
- Define a subspec `common` only with our source code. Then another subspec called `SQLite` than depends on `common` and `FMDB` and a third one called `SQLCipher` that depends on `common` and `FMDB/SQLCipher`. Finally I have to replace `pod lib lint --allow-warnings` with:
```
  sh "pod lib lint --allow-warnings --subspec='CDTDatastore/SQLite'"
  sh "pod lib lint --allow-warnings --subspec='CDTDatastore/SQLCipher'"
```
- Do not define `common` and instead copy its content in `SQLite` and `SQLCipher`. It is not necessary to modify `Rakefile`

indisoluble replied:
Remove `pod lib lint` from Rakefile (or executing it only for a couple of subspecs) does not solve the problem. @mikerhodes mentioned that to release a new version of the library, we have to execute `pod trunk push`; this command runs a local lint of the podspec before pushing it to podspec.

Therefore the only feasible option is the third one.

indisoluble replied:
Fixed in e90ee2b

mikerhodes commented on Rakefile:
```diff
@@ -72,8 +110,13 @@ task :testdevice do
 
+desc "pod update all included projects"
+task :podupdate do
+  sh "rake podupdatetests"
```
I think you can call the `podupdatetests` task without going via `sh`.

indisoluble replied:
Fixed in 1cc58b3

indisoluble commented on Rakefile:
```diff
@@ -32,22 +32,60 @@ task :testosx do
+desc "Run the CDTDatastore Encryption Tests for iOS"
+task :testencryptionios do
+    # build using xcpretty as otherwise it's very verbose when running tests
+  $ios_success = system("xcodebuild -workspace EncryptionTests/EncryptionTests.xcworkspace -scheme 'Encryption Tests' -destination 'platform=iOS Simulator,OS=8.1,name=iPhone 4S' build | xcpretty; exit ${PIPESTATUS[0]}")
```
OS should be `latest`

indisoluble replied:
Fixed in df17a3b

indisoluble commented on Rakefile:
```diff
@@ -32,22 +32,60 @@ task :testosx do
+    puts "** Build failed"
+    exit(-1)
+  end
+  $ios_success = system("xcodebuild -workspace EncryptionTests/EncryptionTests.xcworkspace -scheme 'Encryption Tests' -destination 'platform=iOS Simulator,OS=8.1,name=iPhone 4S' test")
```
As in the previous comment, OS should be `lastest`

indisoluble replied:
Fixed in df17a3b

mikerhodes commented:
A +1 from me, awaiting @emlaver

emlaver commented:
+1

indisoluble commented:
Not merged to master, new branch feature-encryption created to hold whole feature while allowing smaller PRs as development occurs
